GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Weapon.cpp

```diff
@@ -1254,6 +1263,53 @@ void WeaponTemplate::dealDamageInternal(ObjectID sourceID, ObjectID victimID, co
 		}
 
 	}
+}
+#else
+void WeaponTemplate::processHistoricDamage(const Object* source, const Coord3D* pos) const
+{
+	if (m_historicBonusCount > 0 && m_historicBonusWeapon != this)
+	{
+		trimOldHistoricDamage();
+
+		Real radSqr = m_historicBonusRadius * m_historicBonusRadius;
+		Int count = 0;
+		for (HistoricWeaponDamageList::iterator it = m_historicDamage.begin(); it != m_historicDamage.end(); ++it)
+		{
+			if (is2DDistSquaredLessThan(*pos, it->location, radSqr))
+			{
+				// This one is close enough in time and distance, so count it. This is tracked by template since it applies
+				// across units, so don't try to clear historicDamage on success in here.
+				(*it).triggered = true;
+				++count;
+			}
+		}
+
+		if (count >= m_historicBonusCount - 1)	// minus 1 since we include ourselves implicitly
+			TheWeaponStore->createAndFireTempWeapon(m_historicBonusWeapon, source, pos);
+		else
+		{
+			for (HistoricWeaponDamageList::iterator it = m_historicDamage.begin(); it != m_historicDamage.end(); ++it)
+				(*it).triggered = false;
```

Comment:
I think we can get rid of this by having a 

`UnsignedInt HistoricWeaponDamageInfo::triggerFrame`

which is then set to current frame when triggered, and removed when trigger count is confirmed.

Then have 2 functions:

`trimExpiredHistoricDamage()`

and

`trimTriggeredHistoricDamage(triggerFrame)`

Make `trimExpiredHistoricDamage` efficient by expecting chronological order and call it same place as `trimOldHistoricDamage`.

Call `trimTriggeredHistoricDamage` after the count test.